tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,496 @@
+                waker: None,
+                notified: false,
+            }),
+        }
+        .await
+    }
+
+    /// Notifies a waiting task
+    ///
+    /// If a task is currently waiting, that task is notified. Otherwise, a
+    /// permit is stored in this `Notify` value and the **next** call to
+    /// `recv()` will complete immediately consuming the permit made available
+    /// by this call to `notify_one()`.
+    ///
+    /// At most one permit may be stored by `Notify`. Many sequential calls to
+    /// `notify_one` will result in a single permit being stored. The next call
+    /// to `recv()` will complete immediately, but the one after that will wait.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::Notify;
+    /// use std::sync::Arc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let notify = Arc::new(Notify::new());
+    ///     let notify2 = notify.clone();
+    ///
+    ///     tokio::spawn(async move {
+    ///         notify2.recv().await;
+    ///         println!("received a notification");
+    ///     });
+    ///
+    ///     notify.notify_one();
+    /// }
+    /// ```
+    pub fn notify_one(&self) {
+        // Load the current state
+        let mut curr = self.state.load(SeqCst);
+
+        // If the state is `EMPTY`, transition to `NOTIFIED` and return.
+        while let EMPTY | NOTIFIED = curr {
+            // The compare-exchange from `NOTIFIED` -> `NOTIFIED` is
+            // intended a happens-before synchronization must happen
+            // between this atomic operation and a task calling
+            // `recv()`.
+            let res = self.state.compare_exchange(curr, NOTIFIED, SeqCst, SeqCst);
+
+            match res {
+                // No waiters, no further work to do
+                Ok(_) => return,
+                Err(actual) => {
+                    curr = actual;
+                }
+            }
+        }
+
+        // There are waiters, the lock must be acquired to notify.
+        let mut waiters = self.waiters.lock().unwrap();
+
+        // The state must be reloaded while the lock is held. The state may only
+        // transition out of WAITING while the lock is held.
+        curr = self.state.load(SeqCst);
+
+        if let Some(waker) = notify_locked(&mut waiters, &self.state, curr) {
+            drop(waiters);
+            waker.wake();
+        }
+    }
+}
+
+impl Default for Notify {
+    fn default() -> Notify {
+        Notify::new()
+    }
+}
+
+fn notify_locked(
+    waiters: &mut LinkedList<Waiter>,
+    state: &AtomicU8,
+    mut curr: u8,
+) -> Option<Waker> {
+    loop {
+        match curr {
+            EMPTY | NOTIFIED => {
+                let res = state.compare_exchange(curr, NOTIFIED, SeqCst, SeqCst);
+
+                match res {
+                    Ok(_) => return None,
+                    Err(actual) => {
+                        curr = actual;
```

Comment:
Unless I'm reading this code wrong, `actual` _must_ be either `EMPTY` or `NOTIFIED` here. It cannot be `WAITING`, because that is only ever entered under the lock. Since we are going to continue until we succeed here anyway, and we are overwriting with a constant value, is there a reason to to just do a `state.store(NOTIFIED)`, `return None`, and call it a day?

Reply:
`actual` can be `WAITING` at this point. A thread may concurrently store `WAITING` right before this thread acquires the lock.

Reply:
Actually, the state is reloaded with the lock held, so I think you are right.